exec: fix planning of count operator

yuzefovich opened the pull request:
Previously, when planning a count operator, we would add it to the
flow and would ignore any post-operator planning (like projections).
Now, this is fixed.

Additionally, this commit fixes slicing within projections operators -
previously, we would always slice up to BatchSize, but the underlying
memory not always has sufficient capacity (for example, count operator
uses a batch with a capacity of 1) which would cause an index out of
bounds.

Fixes: #38752.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/38767)
<!-- Reviewable:end -->

jordanlewis reviewed (commented):
Fix for COUNT <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

This is underscoring for me that I think it's time to think about refactoring the column exec setup stuff. I think we should move it into a different package and give each case in this giant switch its own function, so that the code doesn't have to "remember" stuff like not being allowed to just return.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38767#-:-LjQwlyV3S-jD8nepPwT:bqtubuh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @solongordon and @yuzefovich)

___
*[pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38767#-LjQyVl13aVjHzxrqhyJ:-LjQyVl13aVjHzxrqhyK:b-hqw5gi) ([raw file](https://github.com/cockroachdb/cockroach/blob/694e9afb54e6656d9c4ac922b04b01caaa92ce3a/pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go#L65)):*
> ```Go
> 	vec := batch.ColVec(p.colIdx)
> 	col := vec.{{.LTyp}}()[:coldata.BatchSize]
> 	projVec := batch.ColVec(p.outputIdx)
> ```

I think the purpose of these slices was to eliminate bounds checks
___
*[pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38767#-LjQydRB9GQLTiYq2oEK:-LjQydRB9GQLTiYq2oEL:bac1cun) ([raw file](https://github.com/cockroachdb/cockroach/blob/9361947a05cc4e89cfaf7679f7cc245b28ef50b9/pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go#L73)):*
> ```Go
> 	} else {
> 		col := vec.{{.LTyp}}()[:n]
> 		projCol := projVec.{{.RetTyp}}()[:n]
> ```

I think the purpose of the len(col)-1 thing was to eliminate bounds checks.
___
*[pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go, line 110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38767#-LjQylDA8THve2o9NXx_:-LjQylDA8THve2o9NXxa:bzqspa) ([raw file](https://github.com/cockroachdb/cockroach/blob/9361947a05cc4e89cfaf7679f7cc245b28ef50b9/pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go#L110)):*
> ```Go
> 	projVec := batch.ColVec(p.outputIdx)
> 	if sel := batch.Selection(); sel != nil {
> 		col := vec.{{.RTyp}}()[:coldata.BatchSize]
> ```

I'm confused why it's okay to do this slicing up to batch size in the case of sel?


<!-- Sent from Reviewable.io -->

yuzefovich committed:
exec: fix planning of count operator

Previously, when planning a count operator, we would add it to the
flow and would ignore any post-operator planning (like projections).
Now, this is fixed.

Additionally, this commit fixes slicing within projections operators -
previously, we would always slice up to BatchSize, but the underlying
memory not always has sufficient capacity (for example, count operator
uses a batch with a capacity of 1) which would cause an index out of
bounds.

Release note: None

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38767)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis and @solongordon)

___
*[pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38767#-LjQyVl13aVjHzxrqhyJ:-LjS1NxJ-ZTATqk_mKws:b-wn2o95) ([raw file](https://github.com/cockroachdb/cockroach/blob/694e9afb54e6656d9c4ac922b04b01caaa92ce3a/pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go#L65)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think the purpose of these slices was to eliminate bounds checks
</blockquote></details>

Yeah,  I knew that. I took a look at compiled code and didn't see a noticeable change. But I was wrong - I should've used `check_bce` flag and look at the number lines as Solon suggested. Now, I've checked the three places where I'm making modifications, and bounds checks are eliminated in all of them.
___
*[pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38767#-LjQydRB9GQLTiYq2oEK:-LjS1muX3LlLiGTsB6qb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9361947a05cc4e89cfaf7679f7cc245b28ef50b9/pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go#L73)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think the purpose of the len(col)-1 thing was to eliminate bounds checks.
</blockquote></details>

Done.
___
*[pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go, line 110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38767#-LjQylDA8THve2o9NXx_:-LjS1szj8kFZ99ZMD9hV:b-b4rq3k) ([raw file](https://github.com/cockroachdb/cockroach/blob/9361947a05cc4e89cfaf7679f7cc245b28ef50b9/pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go#L110)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I'm confused why it's okay to do this slicing up to batch size in the case of sel?
</blockquote></details>

Well, it's not really ok if the underlying batch does not have the required capacity. I guess count operator is the only one (?) that uses a batch with not 1024 capacity, and if we put a filter on top of it and then a projection, it'll fail.

I checked, and top k also uses a custom capacity.

Selections were also prone to this problem. I'm not sure whether slicing up to capacity is the right thing here though. Thoughts?


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Guys, please take another look.

solongordon reviewed (changes requested):
Good catch that we should not be arbitrarily slicing up to `BatchSize`. However I'm not sure that the new bounds check elimination stuff you added is really doing anything. See my inline comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38767#-:-LjvC2Le7zs14apCMtaC:b-9ar3yi)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @solongordon, and @yuzefovich)

___
*[pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go, line 72 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38767#-LjvAmOI3iMvhwPsA-Tz:-LjvAmOI3iMvhwPsA-U-:b-izo3nh) ([raw file](https://github.com/cockroachdb/cockroach/blob/53753018c27c466b92ede73496776b1fec177ce0/pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go#L72)):*
> ```Go
> 		} else {
> 			col = col[:coldata.BatchSize]
> 		}
> ```

I'm not convinced there's any bounds check being eliminated here. It seems like this would only eliminate a bounds check if the compiler knew that `i` in the loop below is less than `cap(col)` and `coldata.BatchSize`, which it cannot. I tried getting rid of these 5 lines and did not see any more bounds checks when I did `check_bce`. Same goes for the other selection vector cases.

I don't see a way to do BCE here, so I would just get rid of the slicing entirely.


<!-- Sent from Reviewable.io -->

solongordon reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38767)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @solongordon, and @yuzefovich)

___
*[pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go, line 65 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38767#-LjvCs5UDCU3Nfi2uWuz:-LjvCs5UDCU3Nfi2uWv-:b-ugpzd6) ([raw file](https://github.com/cockroachdb/cockroach/blob/53753018c27c466b92ede73496776b1fec177ce0/pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go#L65)):*
> ```Go
> 	vec := batch.ColVec(p.colIdx)
> 	projVec := batch.ColVec(p.outputIdx)
> 	projCol := projVec.{{.RetTyp}}()[:coldata.BatchSize]
> ```

While we're at it I think this `[:coldata.BatchSize]` slice is also pointless. (Same goes for the other one below.)


<!-- Sent from Reviewable.io -->

solongordon reviewed (commented):
BTW, I think `select_in_tmpl.go` has the same `BatchSize` issue if you'd like to fix that here too.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38767#-:-LjvDIGxEPjY7gpUJBhd:bgj8a9n)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @solongordon, and @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Hm, I didn't even know about `select_in_tmpl` :)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38767#-:-LjvUtWU8fVHam6FqlO5:b-y28swm)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis and @solongordon)

___
*[pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go, line 65 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38767#-LjvCs5UDCU3Nfi2uWuz:-LjvTBPZ91PlCv_BKdV3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/53753018c27c466b92ede73496776b1fec177ce0/pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go#L65)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

While we're at it I think this `[:coldata.BatchSize]` slice is also pointless. (Same goes for the other one below.)
</blockquote></details>

Done.
___
*[pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go, line 72 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38767#-LjvAmOI3iMvhwPsA-Tz:-LjvTCwMCEGcpGADKGxZ:b-nzcdio) ([raw file](https://github.com/cockroachdb/cockroach/blob/53753018c27c466b92ede73496776b1fec177ce0/pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go#L72)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I'm not convinced there's any bounds check being eliminated here. It seems like this would only eliminate a bounds check if the compiler knew that `i` in the loop below is less than `cap(col)` and `coldata.BatchSize`, which it cannot. I tried getting rid of these 5 lines and did not see any more bounds checks when I did `check_bce`. Same goes for the other selection vector cases.

I don't see a way to do BCE here, so I would just get rid of the slicing entirely.
</blockquote></details>

Makes sense, removed.


<!-- Sent from Reviewable.io -->

solongordon reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I think the `BatchSize` stuff goes back to some cargo culting on my part before I understood how BCE works, so thanks for cleaning up my mess. :)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38767#-:-LjvW8sSBbAKrR3ZSRuZ:b1zhu46)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
Thanks for the review! I was glad to take a look at this stuff to learn more about BCE myself :)

bors r+
